File: laocai-bot-spring-boot-starter/src/test/java/indi/kyson/laocai/bot/model/segment/SegmentTest.java
package indi.kyson.laocai.bot.model.segment;

import org.junit.jupiter.api.Test;
import tools.jackson.databind.json.JsonMapper;

import static org.assertj.core.api.Assertions.assertThat;

class SegmentTest {

    private final JsonMapper jsonMapper = JsonMapper.builder().build();

    @Test
    void textSegmentSerializesToTypeAndDataShape() {
        Segment segment = TextSegment.of("hi");

        var tree = jsonMapper.valueToTree(segment);

        assertThat(tree.get("type").asString()).isEqualTo("text");
        assertThat(tree.get("data").get("text").asString()).isEqualTo("hi");
        assertThat(tree.size()).isEqualTo(2);
    }

    @Test
    void textSegmentRoundTripsThroughDeserialization() {
        Segment segment = jsonMapper.readValue("{\"type\":\"text\",\"data\":{\"text\":\"hi\"}}", Segment.class);

        assertThat(segment).isInstanceOf(TextSegment.class);
        assertThat(((TextSegment) segment).getText()).isEqualTo("hi");
    }

    @Test
    void unknownTypeFallsBackToUnknownSegmentInsteadOfThrowing() {
        Segment segment = jsonMapper.readValue("{\"type\":\"foobar\",\"data\":{\"x\":1}}", Segment.class);

        assertThat(segment).isInstanceOf(UnknownSegment.class);
        assertThat(segment.getType()).isEqualTo("foobar");
    }
}

File: laocai-bot-core/src/test/kotlin/indi/kyson/laocai/bot/core/segment/SegmentTest.kt
package indi.kyson.laocai.bot.core.segment

import tools.jackson.databind.JsonNode
import tools.jackson.databind.json.JsonMapper
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs

class SegmentTest {

    private val jsonMapper = JsonMapper.builder().build()

    @Test
    fun textSegmentSerializesToTypeAndDataShape() {
        val segment: Segment = TextSegment.of("hi")

        val tree = jsonMapper.valueToTree<JsonNode>(segment)

        assertEquals("text", tree.get("type").asString())
        assertEquals("hi", tree.get("data").get("text").asString())
        assertEquals(2, tree.size())
    }

    @Test
    fun textSegmentRoundTripsThroughDeserialization() {
        val segment = jsonMapper.readValue("{\"type\":\"text\",\"data\":{\"text\":\"hi\"}}", Segment::class.java)

        assertIs<TextSegment>(segment)
        assertEquals("hi", segment.text)
    }

    @Test
    fun unknownTypeFallsBackToUnknownSegmentInsteadOfThrowing() {
        val segment = jsonMapper.readValue("{\"type\":\"foobar\",\"data\":{\"x\":1}}", Segment::class.java)

        assertIs<UnknownSegment>(segment)
        assertEquals("foobar", segment.type)
    }
}
